pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,231 @@
+from torch.utils.data.dataloader import IterableDataset
+
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler(Sampler):
+    """
+    This class is used to wrap a :class:`torch.utils.data.Sampler` and record the number
+    of iteration performed during an epoch.
+    On reload, if a ``state_dict`` is provided, this will be used to fast forward the wrapped samplers.
+
+    """
+
+    def __init__(self, sampler: Union[Sampler, BatchSampler, Generator]) -> None:
+        self._sampler = sampler
+        self._current_iteration = 0
+        self._dataloader_batch_size: Optional[int] = None
+        self.restarting: bool = False
+        self._cached_state_dict: Optional[Dict[str, Any]] = None
+
+    def setup(self, dataloader_batch_size: Optional[int] = None) -> None:
+        """
+        Setup the ``FastForwardSampler``.
+        This is required only when the provided dataset subclassed :class:`torch.utils.data.Dataset`.
+        """
+        self._dataloader_batch_size = dataloader_batch_size
+
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    def __iter__(self) -> Iterator[List[int]]:
+        # setup local counter
+        restart_counter = 0
+
+        # iteration over wrapped sampler
+        for batch in self._sampler:
+
+            # if we are restarting, we will fastforward the batches
+            if self.restarting:
+
+                # the state dict is cached until workers are made available through the DataLoader
+                if self._cached_state_dict is not None and self.worker_id in self._cached_state_dict:
+
+                    # reload the current state dict
+                    self.load_state_dict(self._cached_state_dict, workers_initialized=True)
+                    self._cached_state_dict = None
+
+                # increment counter
+                restart_counter += 1
+
+                # if restart counter matches the current iteration, we should stop restarting
+                if restart_counter == self._current_iteration:
+                    self.restarting = False
+            else:
+                self._current_iteration += 1
+
+                # yield the batch
+                yield batch
+
+        self.reset()
+
+    def reset(self) -> None:
+        self._current_iteration = 0
+
+    def __len__(self) -> int:
+        return len(self._sampler)
+
+    @property
+    def drop_last(self) -> bool:
+        return self._sampler.drop_last
+
+    @property
+    def batch_size(self) -> int:
+        return self._sampler.batch_size
+
+    @property
+    def sampler(self) -> Sampler:
+        return self._sampler
+
+    @property
+    def batch_indices(self) -> Optional[List[int]]:
+        return self._sampler.batch_indices
+
+    def _compute_current_iteration(self, number_batch_processed: Optional[int] = None) -> int:
```

Comment:
```suggestion
    def _compute_current_iteration(self, num_batches_processed: Optional[int] = None) -> int:
```
Nit: How about this?

Reply:
Done !